src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -357,14 +223,375 @@ public void enable(boolean enabled)
         metrics.reset();
     }
 
+    public boolean isEnabled()
+    {
+        return enabled;
+    }
+
     @VisibleForTesting
     public void intercept(Function<RebuffererFactory, RebuffererFactory> interceptor)
     {
         final Function<ChunkReader, RebuffererFactory> prevWrapper = wrapper;
         wrapper = rdr -> interceptor.apply(prevWrapper.apply(rdr));
     }
 
-    // TODO: Invalidate caches for obsoleted/MOVED_START tables?
+    /**
+     * Maps a reader to a reader id, used by the cache to find content.
+     *
+     * Uses the file name (through the fileIdMap), reader type and chunk size to define the id.
+     * The lowest {@link #READER_TYPE_BITS} are occupied by reader type, then the next {@link #CHUNK_SIZE_LOG2_BITS}
+     * are occupied by log 2 of chunk size (we assume the chunk size is the power of 2), and the rest of the bits
+     * are occupied by fileId counter which is incremented for each unseen file name.
+     */
+    private long readerIdFor(File file, ChunkReader.ReaderType type, int chunkSize)
+    {
+        return (((fileIdMap.computeIfAbsent(file, this::assignFileId)
+                  << CHUNK_SIZE_LOG2_BITS) | Integer.numberOfTrailingZeros(chunkSize))
+                << READER_TYPE_BITS) | type.ordinal();
+    }
+
+    /**
+     * Maps a reader to a file id, used by the cache to find content.
+     */
+    protected long readerIdFor(ChunkReader source)
+    {
+        return readerIdFor(source.channel().getFile(), source.type(), source.chunkSize());
+    }
+
+    private long assignFileId(File file)
+    {
+        return nextFileId.getAndIncrement();
+    }
+
+    /**
+     * Invalidate all buffers from the given file, i.e. make sure they can not be accessed by any reader using a
+     * FileHandle opened after this call. The buffers themselves will remain in the cache until they get normally
+     * evicted, because it is too costly to remove them.
+     *
+     * Note that this call has no effect of handles that are already opened. The correct usage is to call this when
+     * a file is deleted, or when a file is created for writing. It cannot be used to update and resynchronize the
+     * cached view of an existing file.
+     */
+    public void invalidateFile(File file)
+    {
+        // Removing the name from the id map suffices -- the next time someone wants to read this file, it will get
+        // assigned a fresh id.
+        fileIdMap.remove(file);
+    }
+
+    /**
+     * Removes FileId for given file if cache is initialized.
+     * @param file file to remove from the map.
+     */
+    public static void removeFileIdFromCache(File file)
+    {
+        if (instance != null)
```

Comment:
we may have multiple instances of the ChunkCache, in CNDB we have the IndexChunkCache
I see that this method is used only in the SequentialWriter class, what about adding some comment or improve the javadocs ?

Reply:
I intend to move the invalidation to `FileHandle.Builder`, or move it back to the tidier.

This will take some time, though, I need to write new tests to make sure we are not releasing content we shouldn't.